chefkoch/recipe.py

```diff
@@ -0,0 +1,771 @@
+        """
+        is_unicode = False
+        try:
+            is_unicode = isinstance(name, unicode)
+        except NameError as mimimi:
+            logger.debug(mimimi)
+            logger.debug("You are using python 3, but don't worry," +
+                         "we make it work.")
+            pass
+        if not (isinstance(name, str) or is_unicode):
+            raise TypeError('The name of a node must be a string.')
+        if not self.is_ascii(name):
+            raise ValueError('The name of a node must be ascii.')
+        self.name = name
+    
+    def is_ascii(self, name):
+        """
+        Checks if string consists of only ascii characters.
+
+        :param name: string
+        :type name: string or unicode
+        :return: 
+            `True`, if name only contains ascii characters.
+        """
+        return all(ord(c) < 128 for c in name)
+
+
+class StepSource:
+    """
+    Specifies the function to be executed inside a node in the recipe.
+    """
+    def __init__(self, stepsource):
+        """
+        Tests the step source if it is a recipe, a python executable or
+        a built-in function and initialises it if so.
+
+        :param stepsource: file path to a sub-recipe, a python executable
+            or the name of a built-in function
+        :type stepsource: string
+        :raises TypeError: If the string does not match any of the above.
+        """
+        # testing if step is built-in; JSON file or python function
+        extension = os.path.splitext(stepsource)[1]
+        if extension == ".py":
+            self.step = stepsource
+        elif extension == ".json":
+            self.step = stepsource
+        elif str(stepsource) in BUILT_INS:
+            self.step = stepsource
+            # done: research on assigning functions as attributes
+            # (so that it can be accessed no matter where the object
+            # is used)
+        else:
+            raise TypeError('Stepsource to node ' + str(name) +
+                            ': ' + str(stepsource) +
+                            '. Must be a Python file, another recipe ' +
+                            'or a build-in function.')
+
+
+class Flavour(dict):
+    """
+    The Flavour class extends the dictionary class and holds the parsed flavour
+    file. The flavour file is the collection of all paramters needed for the
+    simulation and all their values the simulation should be executed with. The
+    goal is to find the best parameter combination. Paramter can have a constant
+    value, a list of values or a range. They can also be files.
+    """
+
+    def tostring(self):
+        """
+        Converts the data held inside the flavour object into a string.
+
+        :returns:
+            string - Holds structured content of the flavour object
+        """
+        content = "The flavour file contains: "
+        for key in self:
+            content += "\n  " + self[key].tostring()
```

Comment:
calling `<someobject>.tostring()` requires a method `tostring()` to be present.
The better approach to do this is to resort to the python built-ins `str(myObject)`, which is equivalent to `myObject.__str__(self)` and can be defined exactly in this way.
Please note, that  `myClass.__repr__(self)` partners up with `repr(myObject)` (with `myObject` being an instance of `myClass`)

Eery object implements the two.